lib/src/Pos/Diffusion/Full/Block.hs

```diff
@@ -124,11 +120,27 @@ getBlocks logic enqueue nodeId tipHeader checkpoints = do
     -- then you can skip requesting the headers and go straight to requesting
     -- the block itself.
     blocks <- if singleBlockHeader
-              then requestBlocks (NewestFirst (pure tipHeader))
-              else requestHeaders >>= requestBlocks
+              then requestBlocks (OldestFirst (one tipHeader))
+              else twoPhaseAction
     pure (OldestFirst (reverse (toList blocks)))
   where
 
+    twoPhaseAction = do
```

Comment:
Needs a comment and a better name if you can think of one. How about `getAndClassifyHeaders`, then take out `requestBlocks` from the end and put it back up on line 124?

Reply:
Sure, I will.